main.go

```diff
@@ -95,114 +96,165 @@ func getEtcdConfig() (etcd.Config, error) {
-	if errorButKeyNotFound(err) != nil {
+func (s *Server) getPeerASN(host string) (numorstring.ASNumber, error) {
+	node, err := s.cli.Nodes().Get(calicoapi.NodeMetadata{Name: host})
+	if err != nil {
 		return 0, err
 	}
-	if res != nil && res.Node != nil {
-		v, err := strconv.ParseUint(res.Node.Value, 10, 32)
-		if err != nil {
-			return 0, err
-		}
-		return uint32(v), nil
+	if node.Spec.BGP == nil {
+		return 0, fmt.Errorf("host %s is running in policy-only mode")
 	}
-	return getGlobalASN(api)
+	asn := node.Spec.BGP.ASNumber
+	if asn == nil {
+		return s.getGlobalASN()
+	}
+	return *asn, nil
 
 }
 
-func getGlobalConfig(api etcd.KeysAPI) (*bgpconfig.Global, error) {
-	asn, err := getGlobalASN(api)
+func (s *Server) getGlobalConfig() (*bgpconfig.Global, error) {
+	asn, err := s.getGlobalASN()
 	if err != nil {
 		return nil, err
 	}
 	return &bgpconfig.Global{
 		Config: bgpconfig.GlobalConfig{
-			As:       asn,
+			As:       uint32(asn),
 			RouterId: os.Getenv(IP),
 		},
 	}, nil
 }
 
-func isMeshMode(api etcd.KeysAPI) (bool, error) {
-	res, err := api.Get(context.Background(), fmt.Sprintf("%s/global/node_mesh", CALICO_BGP), nil)
-	if err != nil {
-		return false, err
-	}
-	m := &struct {
-		Enabled bool `json:"enabled"`
-	}{}
-	if err := json.Unmarshal([]byte(res.Node.Value), m); err != nil {
-		return false, err
-	}
-	return m.Enabled, nil
+func (s *Server) isMeshMode() (bool, error) {
+	return s.cli.Config().GetNodeToNodeMesh()
 }
 
-func getMeshNeighborConfigs(api etcd.KeysAPI) ([]*bgpconfig.Neighbor, error) {
-	globalASN, err := getGlobalASN(api)
+func (s *Server) getMeshNeighborConfigs() ([]*bgpconfig.Neighbor, error) {
+	globalASN, err := s.getGlobalASN()
 	if err != nil {
 		return nil, err
 	}
-	res, err := api.Get(context.Background(), fmt.Sprintf("%s/host", CALICO_BGP), &etcd.GetOptions{Recursive: true})
+	nodes, err := s.cli.Nodes().List(calicoapi.NodeMetadata{})
 	if err != nil {
 		return nil, err
 	}
-	ns := make([]*bgpconfig.Neighbor, 0, len(res.Node.Nodes))
-	for _, node := range res.Node.Nodes {
+	ns := make([]*bgpconfig.Neighbor, 0, len(nodes.Items))
+	for _, node := range nodes.Items {
 		var v4, v6 string
 		peerASN := globalASN
-		for _, v := range node.Nodes {
-			path := strings.Split(v.Key, "/")
-			key := path[len(path)-1]
-			switch key {
-			case "ip_addr_v4":
-				v4 = v.Value
-				if v4 == os.Getenv(IP) {
-					v4 = ""
-				}
-			case "ip_addr_v6":
-				v6 = v.Value
-				if v6 == os.Getenv(IP6) {
-					v6 = ""
-				}
-			case "as_num":
-				asn, err := strconv.ParseUint(v.Value, 10, 32)
-				if err != nil {
-					return nil, err
-				}
-				peerASN = uint32(asn)
-			default:
-				log.Printf("unhandled key: %s", v.Key)
+		spec := node.Spec.BGP
+		if spec == nil {
+			continue
+		}
+
+		asn := spec.ASNumber
+		if asn != nil {
+			peerASN = *asn
+		}
+		if spec.IPv4Address != nil {
+			v4 = spec.IPv4Address.String()
+			if v4 == os.Getenv(IP) {
+				v4 = ""
```

Comment:
I'm setting `v4` or `v6` empty to avoid adding the current node as a BGP neighbor.

We don't add a neighbor when either is empty [here](https://github.com/projectcalico/calico-bgp-daemon/pull/14/files#diff-7ddfb3e035b42cd70649cc33393fe32cR242)